src/Polar/Constraints/power_balance.jl

```diff
@@ -75,35 +75,42 @@ function adjoint!(
     )
 end
 
-function matpower_jacobian(polar::PolarForm, ::State, ::typeof(power_balance), V)
-    nbus = get(polar, PS.NumberOfBuses())
-    pf = polar.network
-    ref = pf.ref
-    pv = pf.pv
-    pq = pf.pq
-    Ybus = pf.Ybus
-
-    dSbus_dVm, dSbus_dVa = _matpower_residual_jacobian(V, Ybus)
-    j11 = real(dSbus_dVa[[pv; pq], [pv; pq]])
-    j12 = real(dSbus_dVm[[pv; pq], pq])
-    j21 = imag(dSbus_dVa[pq, [pv; pq]])
-    j22 = imag(dSbus_dVm[pq, pq])
-
-    return [j11 j12; j21 j22]
-end
-function matpower_jacobian(polar::PolarForm, ::Control, ::typeof(power_balance), V)
+function matpower_jacobian(polar::PolarForm, X::Union{State, Control}, ::typeof(power_balance), V)
     nbus = get(polar, PS.NumberOfBuses())
     pf = polar.network
     ref = pf.ref ; nref = length(ref)
     pv = pf.pv ; npv = length(pv)
     pq = pf.pq ; npq = length(pq)
     Ybus = pf.Ybus
 
-    dSbus_dVm, _ = _matpower_residual_jacobian(V, Ybus)
-    j11 = real(dSbus_dVm[[pv; pq], [ref; pv]])
-    j12 = sparse(I, npv + npq, npv)
-    j21 = imag(dSbus_dVm[pq, [ref; pv]])
-    j22 = spzeros(npq, npv)
-    return [j11 -j12; j21 j22]
+    dSbus_dVm, dSbus_dVa = PS.matpower_residual_jacobian(V, Ybus)
+
+    if isa(X, State)
+        j11 = real(dSbus_dVa[[pv; pq], [pv; pq]])
+        j12 = real(dSbus_dVm[[pv; pq], pq])
+        j21 = imag(dSbus_dVa[pq, [pv; pq]])
+        j22 = imag(dSbus_dVm[pq, pq])
+        return [j11 j12; j21 j22]
+    elseif isa(X, Control)
+        j11 = real(dSbus_dVm[[pv; pq], [ref; pv]])
+        j12 = sparse(I, npv + npq, npv)
+        j21 = imag(dSbus_dVm[pq, [ref; pv]])
+        j22 = spzeros(npq, npv)
+        return [j11 -j12; j21 j22]
+    end
+end
```

Comment:
@frapac I actually prefer this version. Cool!

https://medium.com/@lovelydog/4-notes-modules-should-be-deep-ba5671c4288c